File: week4/java/Subsets_78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        subsets(nums, 0, new ArrayList<Integer>(), result);
        return result;
    }
    
    private void subsets(int[] nums, int start, List<Integer> list, List<List<Integer>> result) {
        
        result.add(new ArrayList<>(list));
        if (start >= nums.length) {
            return;
        }
        
        for (int i = start; i < nums.length; i++) {
            list.add(nums[i]);
            subsets(nums, i + 1, list, result);
            list.remove(list.size() - 1);
        }
    }
    
}

File: week4/python/Subsets_78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        result = []
        
        def subset(index: int, cur: List[int]):
            
            result.append(list(cur))
            if index >= len(nums):
                return
            
            for i in range(index, len(nums), 1):
                cur.append(nums[i])
                subset(i + 1, cur)
                cur[:] = cur[:-1]
                
        subset(0, list())
        
        return result
